miner/algo_common.go

```diff
@@ -181,21 +185,40 @@ func applyTransactionWithBlacklist(signer types.Signer, config *params.ChainConf
 }
 
 // commit tx to envDiff
-func (envDiff *environmentDiff) commitTx(tx *types.Transaction, chData chainData) (*types.Receipt, int, error) {
-	header := envDiff.header
-	coinbase := &envDiff.baseEnvironment.coinbase
-	signer := envDiff.baseEnvironment.signer
+func (envDiff *environmentDiff) commitTx(tx *types.Transaction, chData chainData, algoConf algorithmConfig) (*types.Receipt, int, error) {
+	var (
+		header   = envDiff.header
+		coinbase = &envDiff.baseEnvironment.coinbase
+		signer   = envDiff.baseEnvironment.signer
+		root     = envDiff.state.OriginalRoot()
+	)
 
 	gasPrice, err := tx.EffectiveGasTip(header.BaseFee)
 	if err != nil {
 		return nil, shiftTx, err
 	}
 
 	envDiff.state.SetTxContext(tx.Hash(), envDiff.baseEnvironment.tcount+len(envDiff.newTxs))
-
 	receipt, newState, err := applyTransactionWithBlacklist(signer, chData.chainConfig, chData.chain, coinbase,
 		envDiff.gasPool, envDiff.state, header, tx, &header.GasUsed, *chData.chain.GetVMConfig(), chData.blacklist)
 
+	// when drop transaction is enabled:
+	// 1. if there was an error applying the transaction OR
+	// 2. if the transaction receipt status is failed
+	// we don't apply the transaction to the state and we don't add it to the newTxs, return early
+	if algoConf.DropTransactionOnRevert && (err != nil || (receipt != nil && receipt.Status == types.ReceiptStatusFailed)) {
+		// the StateDB for environment diff is already modified at this point, since it gets mutated when passed in to
+		// applyTransactionWithBlacklist, so we need to revert it
+		s, sdbErr := state.New(root, envDiff.state.Database(), chData.chain.Snapshots())
```

Comment:
I don't thing that this will work. If this bundle is in the middle of the block root of the state before it is unknown to the database and snapshots. 

And why original root is used here? By the name of it it looks like it will revert to the state before anything was applied to the state. 

Anyway, if you need to revert state to the OK state after tx failed to apply (nonce errror, etc) it can be done simpler. Just use statedb.Snapshot and revert.

And if you need to discard tx if it was applied successfully (but its probably not needed - see Mateusz comment above) it could be done without discarding statedb too. See how blocklist is implemented.

Reply:
>  If this bundle is in the middle of the block root of the state before it is unknown to the database and snapshots

I don't think this is true because when we apply the intermediate state hash is calculated - I confirmed this in test as well

> Anyway, if you need to revert state to the OK state after tx failed to apply (nonce errror, etc) it can be done simpler. Just use statedb.Snapshot and revert.

I tried this approach initially but it failed unfortunately. Using snapshot revert was not working.

Reply:
@dvush this is error I get with approach `Just use statedb.Snapshot and revert`:

```
panic: revision id 0 cannot be reverted [recovered]
	panic: revision id 0 cannot be reverted
```

Reply:
> final hash is calculated once

You are right, I was looking at pre-byzantium which calculates intermediate hash.

Since `applyTransactionWithBlacklist` already reverts on error, I have simplified the logic. Please un-resolve if there are still concerns.